streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java

```diff
@@ -0,0 +1,471 @@
+            final ProcessId clientId = finalAssignments.keySet().iterator().next();
+            final KafkaStreamsAssignment previousAssignment = finalAssignments.get(clientId);
+            finalAssignments.put(clientId, previousAssignment.withFollowupRebalance(Instant.ofEpochMilli(0)));
+        }
+        return new TaskAssignment(finalAssignments.values());
+    }
+
+    private void optimizeActive(final ApplicationState applicationState,
+                                final AssignmentState assignmentState) {
+        if (mustPreserveActiveTaskAssignment) {
+            return;
+        }
+
+        final Map<ProcessId, KafkaStreamsAssignment> currentAssignments = assignmentState.buildKafkaStreamsAssignments();
+
+        final Set<TaskId> statefulTasks = applicationState.allTasks().stream()
+            .filter(TaskInfo::isStateful)
+            .map(TaskInfo::id)
+            .collect(Collectors.toSet());
+        final Map<ProcessId, KafkaStreamsAssignment> optimizedAssignmentsForStatefulTasks = TaskAssignmentUtils.optimizeRackAwareActiveTasks(
+            applicationState, currentAssignments, new TreeSet<>(statefulTasks));
+
+        final Set<TaskId> statelessTasks = applicationState.allTasks().stream()
+            .filter(task -> !task.isStateful())
+            .map(TaskInfo::id)
+            .collect(Collectors.toSet());
+        final Map<ProcessId, KafkaStreamsAssignment> optimizedAssignmentsForAllTasks = TaskAssignmentUtils.optimizeRackAwareActiveTasks(
+            applicationState, optimizedAssignmentsForStatefulTasks, new TreeSet<>(statelessTasks));
+
+        assignmentState.processOptimizedAssignments(optimizedAssignmentsForAllTasks);
+    }
+
+    private void optimizeStandby(final ApplicationState applicationState, final AssignmentState assignmentState) {
+        if (applicationState.assignmentConfigs().numStandbyReplicas() <= 0) {
+            return;
+        }
+
+        if (mustPreserveActiveTaskAssignment) {
+            return;
+        }
+
+        final Map<ProcessId, KafkaStreamsAssignment> currentAssignments = assignmentState.buildKafkaStreamsAssignments();
+        final Map<ProcessId, KafkaStreamsAssignment> optimizedAssignments = TaskAssignmentUtils.optimizeRackAwareStandbyTasks(
+            applicationState, currentAssignments);
+        assignmentState.processOptimizedAssignments(optimizedAssignments);
+    }
+
+    private static void assignActive(final ApplicationState applicationState,
+                                     final Collection<KafkaStreamsState> clients,
+                                     final AssignmentState assignmentState,
+                                     final boolean mustPreserveActiveTaskAssignment) {
+        final int totalCapacity = computeStreamThreadCount(clients);
+        if (totalCapacity == 0) {
+            throw new IllegalStateException("`totalCapacity` should never be zero.");
+        }
```

Comment:
this is another superfluous check now, we should be able to trust the assignor inputs